core/src/main/java/org/apache/calcite/rex/RexCall.java

```diff
@@ -173,6 +173,17 @@ public SqlOperator getOperator() {
   public RexCall clone(RelDataType type, List<RexNode> operands) {
     return new RexCall(type, op, operands);
   }
+
+  @Override public boolean equals(Object obj) {
+    if (!(obj instanceof RexCall)) {
+      return false;
+    }
+    return obj == this || Objects.equals(toString(), obj.toString());
```

Comment:
Do you mean `toString` should be a long-term solution?
Do you mean this would be reworked in future? (e.g. to avoid string build for the sole purpose of equals)

Reply:
I'm not sure...but we shouldn't use `node.toString()` everywhere when we are comparing RexNode-s...that's even worse...

I think the rational behind using digest instead of real tree comparison is to avoid the recursive comparision/etc - as long as when someone creates digest; remembers this design decision I think we are okay - but; for the record: I wouldn't expect this from a field like "digest"...